### PR TITLE
GnomAD LOEUF

### LINKED ISSUE #1022: gnomAD PLi
New pLI scores based on the entire gnomAD 2.1 are now available! https://macarthurlab.org/2018/10/17/gnomad-v2-1/
Should replace old exac pLI table!

### LINKED ISSUE #1571: Show GnomAD ratio to expected and Z-score for missense variation
From GMCK-RD meet up. Are these values present in the exac/gnomad pLi file by any chance?

### LINKED ISSUE #4293: Updated GnomAD gene constraints (v4) available
See https://gnomad.broadinstitute.org/help/constraint for discussion and caveats. Most prominent take home from the blog: do display the LOEUF values as they are arguably more useful. 

Files at e.g. https://storage.googleapis.com/gcp-public-data--gnomad/release/v4.0/constraint/gnomad.v4.0.constraint_metrics.tsv
(see https://gnomad.broadinstitute.org/downloads for mirrors).

This will also solve #1022.

### PR DESCRIPTION
**Pull request by dnil**
This PR adds a functionality.

- Fix #1022 
- Fix #1571 
- Fix #4293

Download new GnomAD `tsv` file:
Screenshot 2024-02-08 at 11 04 45

Parse and display values on genes:
Screenshot 2024-02-08 at 11 46 46

The build gene functions look like things ripe to wrap into a class, but let's keep things simpler and do that in another PR. The constraints get a fake-object-style constant here.

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the Pax software available at [https://pax.scilifelab.se/](https://pax.scilifelab.se). The resource you are going to call dibs on is `scout-stage` and the server is `cg-vm1`.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, repeat procedure at [https://pax.scilifelab.se/](https://pax.scilifelab.se), which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

<details>
<summary>Testing on hasta server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. `ssh <USER.NAME>@hasta.scilifelab.se`
1. Book your testing time using the Pax software. `us; paxa -u <user> -s hasta -r scout-stage`. You can also use the WSGI Pax app available at [https://pax.scilifelab.se/](https://pax.scilifelab.se).
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `conda activate S_scout; pip freeze | grep scout-browser`
1. Deploy the branch to test: `bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_scout -t scout -b <this_branch>`
1. Make sure the branch is deployed: `us; scout --version`
1. After testing is done, repeat the `paxa` procedure, which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

**How to test**:
1. how to test it, possibly with real cases/data

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [ ] code approved by
- [ ] tests executed by

### CODE REVIEW

**Comment by dnil** [comment]
- [ ] Add a reduced demo file of the GnomAD format file.

**Comment by northwestwitch** [comment]
LOEUF:

image

**Comment by northwestwitch** [comment]
I have a stupid question: why do we parse and save these values in the genes when we could simply link out to gnomAD instead?

**Comment by dnil** [comment]
Well, why do we store any external db derived values.. To be fair this one is a bit more just for display. The initial idea was to use these values more actively for ranking/filtering etc. The better question is perhaps why we don't display them directly on the variant page?

File: scout/commands/download/exac.py [hunk 1]
```diff
@@ -4,31 +4,31 @@
...
+    file_path = out_dir / GNOMAD_CONSTRAINT_FILENAME
+    LOG.info("Download GnomAD constraint info to %s", file_path)
     with file_path.open("w", encoding="utf-8") as outfile:
-        for line in fetch_exac_constraint():
+        for line in fetch_constraint():
             outfile.write(line + "\n")
 
 
-@click.command("exac", help="Download a file with ExAC gene info")
+@click.command("exac", help="Download a file with ExAC/GnomAD constraint info")
```
Comment (northwestwitch) [inline scout/commands/download/exac.py:26]: kindof, but 👍🏻

File: scout/parse/exac.py [hunk 2]
```diff
@@ -1,48 +1,60 @@
...
-    logger.info("Parsing exac genes...")
+    logger.info("Parsing GnomAD constraint...")
     for index, line in enumerate(lines):
         if index == 0:
             header = line.rstrip().split("\t")
         elif len(line) > 10:
-            exac_gene = parse_exac_line(line, header)
-
-            yield exac_gene
+            exac_gene = parse_constraint_line(line, header)
```
Comment (northwestwitch) [inline scout/parse/exac.py:58]: Perhaps rename this variable to reflect the new data?

File: scout/parse/exac.py [hunk 3]
```diff
@@ -1,48 +1,60 @@
...
-    Returns:
-        exac_info(dict): A dictionary with the relevant info
+def parse_constraint_line(line: str, header: list) -> Optional[dict]:
+    """Parse a GnomAD constraint formatted line
+
+    In contrast to old ExAC constraint files, these have one line per transcript.
+    We keep only the MANE select ones as representative for the gene, and will in practice only retain one (there will commonly be
+    at least one ENSEMBL and one RefSeq) as these are later aggregated per gene.
     """
     exac_gene = {}
```
Comment (northwestwitch) [inline scout/parse/exac.py:24]: rename it to something else that applies?

File: scout/utils/scout_requests.py [hunk 4]
```diff
@@ -322,35 +322,30 @@ def fetch_hgnc():
     return hgnc_lines
 
 
-def fetch_exac_constraint():
-    """Fetch the file with exac constraint scores
+def fetch_constraint():
```
Comment (northwestwitch) [inline scout/utils/scout_requests.py:325]:
```suggestion
def fetch_gnomad_constraint():
```

no?

File: tests/conftest.py [hunk 5]
```diff
@@ -692,7 +689,6 @@ def parsed_institute(request):
 
 @pytest.fixture(scope="function")
 def institute_obj(request, parsed_institute):
-    print("")
```
Comment (northwestwitch) [inline tests/conftest.py:695]: 🥇

File: tests/conftest.py [hunk 6]
```diff
@@ -1614,7 +1610,7 @@ def panel_handle(request, panel_1_file):
 
 @pytest.fixture
 def hgnc_handle(request, hgnc_file):
-    """Get a file handle to a hgnc file"""
+    """Get a file handle to an hgnc file"""
```
Comment (northwestwitch) [inline tests/conftest.py:1613]: 🥇

**Review by northwestwitch — APPROVED** [review]
Looks good and works fine (tested locally, bot setup demo and downloading the file)

I have a very few comments, great job (3 issues with one PR!)

File: scout/utils/scout_requests.py (hunk 4 above)
Comment (dnil) [reply scout/utils/scout_requests.py:325]: Agree, correct, but if it had been named only "fetch_constraint" from the start I would not have had to rename it.. 😁

**Pull request merged**